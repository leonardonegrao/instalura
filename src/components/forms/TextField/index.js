import React from 'react';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import Text from '../../foundation/Text';

const InputWrapper = styled.div`
  margin-bottom: 17px;
`;

const Input = styled(Text)`
  width: 100%;
  border: 1px solid ${({ theme }) => theme.colors.tertiary.light.color};
  padding: 12px 16px;
  outline: 0;
  border-radius: ${({ theme }) => theme.borderRadius};
`;

Input.defaultProps = {
  tag: 'input',
  variant: 'paragraph1',
};

export default function TextField({
  placeholder,
  name,
  onChange,
  value,
  type = 'text',
  required = false,
}) {
  return (
    <InputWrapper>
      <Input
        type={type}
        placeholder={placeholder}
        name={name}
        onChange={onChange}
        value={value}
        required={required}
      />
    </InputWrapper>
  );
}

TextField.defaultProps = {
  type: 'text',
  required: false,
};

TextField.propTypes = {
  placeholder: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string.isRequired,
  type: PropTypes.string,
  required: PropTypes.bool,
};
