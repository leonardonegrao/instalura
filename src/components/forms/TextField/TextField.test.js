import React from 'react';
import user from '@testing-library/user-event';
import { render, screen } from '../../../infra/tests/testUtils';

import TextField from './index';

describe('<TextField />', () => {
  test('renders component', () => {
    render(
      <TextField
        placeholder="Nome"
        value="Mario"
        onChange={() => {}}
        name="nome"
      />,
    );

    const textField = screen.getByPlaceholderText(/nome/i);

    expect(textField).toMatchSnapshot();
  });
});

describe('when field is valid', () => {
  describe('and user is typing', () => {
    test('the value must be updated', () => {
      const onChangeMock = jest.fn();

      render(
        <TextField
          placeholder="Usu치rio"
          value=""
          onChange={onChangeMock}
          name="user"
        />,
      );

      const textField = screen.getByPlaceholderText(/usu치rio/i);

      const value = 'omariosouto';
      const valueLength = value.length;
      user.type(textField, value);

      expect(onChangeMock).toHaveBeenCalledTimes(valueLength);
    });
  });
});

describe('when field is invalid', () => {
  test('displays the respective error message', () => {
    render(
      <TextField
        placeholder="Email"
        value="devsoutinhogmail.com"
        onChange={() => {}}
        name="email"
        isTouched
        error="O campo email deve conter um e-mail v치lido"
      />,
    );

    const textField = screen.getByPlaceholderText(/email/i);
    const errorSpan = screen.getByRole('alert');

    expect(textField).toHaveValue('devsoutinhogmail.com');
    expect(errorSpan).toHaveTextContent('O campo email deve conter um e-mail v치lido');
    expect(textField).toMatchSnapshot();
  });
});
