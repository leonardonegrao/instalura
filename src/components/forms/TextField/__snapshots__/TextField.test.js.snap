// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TextField /> renders component 1`] = `
.c0 {
  font-family: 'Rubik',sans-serif;
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
}

.c1 {
  width: 100%;
  border: 1px solid #88989E;
  padding: 12px 16px;
  outline: 0;
  border-radius: 12px;
}

<input
  class="c0 c1"
  name="nome"
  placeholder="Nome"
  type="text"
  value="Mario"
/>
`;

exports[`when field is invalid displays the respective error message 1`] = `
.c0 {
  font-family: 'Rubik',sans-serif;
  font-size: 16px;
  font-weight: 400;
  line-height: 1.25;
}

.c1 {
  width: 100%;
  border: 1px solid #88989E;
  padding: 12px 16px;
  outline: 0;
  border-radius: 12px;
  border-color: #dc3545;
}

.c1 + span {
  color: #dc3545;
  font-size: 11px;
}

<input
  class="c0 c1"
  name="email"
  placeholder="Email"
  type="text"
  value="devsoutinhogmail.com"
/>
`;
